### PR TITLE
Make FI usable as a module

### PR DESCRIPTION
**Pull request by g4vroche**
This PR makes it possible to use this example as a node module and extend it.

2 main changes:
 - Split the app initialisation in two file, allowing to change configuration before express server is actually started.
 - Moved the code to an  `src/` folder and created a `lib/` folder for transpiled code
   (run `npm run build` to update builded files.)

Thoses changes should not be breaking for automated deployment tools (app still starts on `/bin/www`)

### CODE REVIEW

**Comment by g4vroche** [comment]
Maybe we should "gitignore" the content of `lib/` folder and publish to npm (with a forced build).
Not sure what the best practice is.

File: bin/www [hunk 1]
```diff
@@ -8,4 +8,4 @@ require('babel-polyfill');
 require('babel-register');
 
 // Import the rest of our application.
-module.exports = require('../app.js');
+module.exports = require('../src/app.js');
```
Comment (rdubigny) [inline bin/www]: the server will not use the compiled code in production ?

File: src/app.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import app from './server';
```
Comment (rdubigny) [inline src/app.js]: this file, should be merged in src/bin/www

File: src/server.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+server.use(helmet());
+server.use(logger('dev'));
+server.use(cookieParser());
+server.use('/assets', express.static(path.join(__dirname, 'public')));
+
+const provider = new Provider(ISSUER, providerConfiguration);
+
+server.provider = provider;
+
+server.start = function start() {
```
Comment (rdubigny) [inline src/server.js]: no need to name the function here, we can directly use : 

```
server.start = () => {
```

**Comment by rdubigny** [comment]
@g4vroche when importing the server.js, it's start connection to database. Is that intended ?

**Comment by g4vroche** [comment]
> @g4vroche when importing the server.js, it's start connection to database. Is that intended ?

Database service is not the part that I was aiming to be able to override, so yes.
But I admit it coud be nice to have the ability to override the database as well as the OIDC provider.

File: src/server.js (hunk 3 above)
Comment (g4vroche) [reply src/server.js]: This was introduced to please the linter ([func-name rule](https://eslint.org/docs/rules/func-names)).
Should we disable this rule in eslint configuration?

File: bin/www (hunk 1 above)
Comment (g4vroche) [reply bin/www]: It definitely should :+1:

File: src/app.js (hunk 2 above)
Comment (g4vroche) [reply src/app.js]: IMO purpose of `./bin/www` is to proxy source code to babel and should not do anything more, while `app.js` is the app launch it self.

File: .huskyrc [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+  "hooks": {
+    "pre-commit": "npm run lint",
+    "pre-push": "npm test"
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (g4vroche) [reply .huskyrc]: This is Atom default behavior AFAIK (and the empty last line is present throught the project)

**Comment by g4vroche** [comment]
> can we make the build a prepublish hook ?

Done and working :rocket: https://www.npmjs.com/package/fi-example

File: src/server.js (hunk 3 above)
Comment (rdubigny) [reply src/server.js]: my bad, I prefer to stick to airbnb configuration.

**Pull request merged**